src/instance_info.rs

```diff
@@ -0,0 +1,191 @@
+use crate::{
+	config::{Config, CONFIG},
+	server::RequestExt,
+	utils::{ErrorTemplate, Preferences},
+};
+use askama::Template;
+use build_html::{Container, Html, HtmlContainer, Table};
+use hyper::{http::Error, Body, Request, Response};
+use once_cell::sync::Lazy;
+use serde::{Deserialize, Serialize};
+use time::OffsetDateTime;
+
+// This is the local static that is intialized at runtime (technically at
+// the first request to the info endpoint) and contains the data
+// retrieved from the info endpoint.
+pub(crate) static INSTANCE_INFO: Lazy<InstanceInfo> = Lazy::new(InstanceInfo::new);
+
+/// Handles instance info endpoint
+pub async fn instance_info(req: Request<Body>) -> Result<Response<Body>, String> {
+	// This will retrieve the extension given, or create a new string - which will
+	// simply become the last option, an HTML page.
+	let extension = req.param("extension").unwrap_or(String::new());
+	let response = match extension.as_str() {
+		"yaml" | "yml" => info_yaml(),
+		"txt" => info_txt(),
+		"json" => info_json(),
+		"html" | "" => info_html(req),
+		_ => {
+			let error = ErrorTemplate {
+				msg: "Error: Invalid info extension".into(),
+				prefs: Preferences::new(&req),
+				url: req.uri().to_string(),
+			}
+			.render()
+			.unwrap();
+			Response::builder().status(404).header("content-type", "text/html; charset=utf-8").body(error.into())
+		}
+	};
+	response.map_err(|err| format!("{err}"))
+}
+
+fn info_json() -> Result<Response<Body>, Error> {
+	let body = serde_json::to_string(&*INSTANCE_INFO).unwrap_or("Error serializing JSON.".into());
+	Response::builder().status(200).header("content-type", "application/json").body(body.into())
+}
+
+fn info_yaml() -> Result<Response<Body>, Error> {
+	let body = serde_yaml::to_string(&*INSTANCE_INFO).unwrap_or("Error serializing YAML.".into());
```

Comment:
See my comment above regarding JSON serialization and the incorrect HTTP status code.